File: addons/lavos/plugins/firebase/crashlytics/FirebaseCrashlytics.cs
using Godot;
using Lavos.Services.Crash;
using System;

namespace Lavos.Plugins.Firebase.Crashlytics;

sealed class FirebaseCrashlytics : ICrashService
{
    const string PluginName = "FirebaseCrashlytics";
    readonly LavosPlugin _plugin;

    public FirebaseCrashlytics()
    {
        Assert.IsTrue(Engine.HasSingleton(PluginName), $"Missing plugins {PluginName}");
        _plugin = new LavosPlugin(Engine.GetSingleton(PluginName));
    }

    public void Initialise()
    {
        _plugin.CallVoid("init");
    }

    #region ICrashService

    public void EnableCollection(bool enable)
    {
        _plugin.CallVoid("setCrashlyticsCollectionEnabled", enable);
    }

    public bool CheckForUnsentReports()
    {
        return _plugin.CallBool("checkForUnsentReports");
    }

    public void SendUnsentReports()
    {
        _plugin.CallVoid("sendUnsentReports");
    }

    public void DeleteUnsentReports()
    {
        _plugin.CallVoid("deleteUnsentReports");
    }

    public bool DidCrashOnPreviousExecution()
    {
        return _plugin.CallBool("didCrashOnPreviousExecution");
    }

    public void Log(string message)
    {
        _plugin.CallVoid("log", message);
    }

    public void LogException(Exception e)
    {
        var exception = $"{e.GetType()} | {e.Message}";
        _plugin.CallVoid("recordException", exception);
    }

    public void SetUserId(string id)
    {
        _plugin.CallVoid("setUserId", id);
    }

    public void SetCustomKey(string key, Variant value)
    {
        var method = string.Empty;
        switch (value.VariantType)
        {
            case Variant.Type.Int: _plugin.CallVoid("setCustomKeyI", key, value); return;
            case Variant.Type.Float: _plugin.CallVoid("setCustomKeyF", key, value); return;
            case Variant.Type.Bool: _plugin.CallVoid("setCustomKeyB", key, value); return;
            case Variant.Type.String: _plugin.CallVoid("setCustomKeyS", key, value); return;
            default:
                {
                    _plugin.CallVoid("setCustomKeyS", key, "TypeNotSupported");
                    Console.Log.Error(nameof(FirebaseCrashlytics), "Unhandled custom type");
                    return;
                }
        }
    }

    public void NativeCrash()
    {
#if DEBUG
        _plugin.CallVoid("crash");
#endif
    }

    #endregion ICrashService
}

File: plugins/android/project/FirebaseCrashlytics/src/main/java/eu/rookies/google/firebase/crashlytics/FirebaseCrashlytics.kt
package eu.rookies.google.firebase.crashlytics

import android.app.Activity
import android.view.View
import androidx.annotation.NonNull
import androidx.annotation.Nullable
import com.google.firebase.FirebaseApp
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.firebase.crashlytics.ktx.crashlytics
import com.google.firebase.ktx.Firebase
import org.godotengine.godot.Godot
import org.godotengine.godot.plugin.GodotPlugin
import org.godotengine.godot.plugin.UsedByGodot

class FirebaseCrashlytics(godot: Godot) : GodotPlugin(godot) {
    private val pluginName = BuildConfig.GODOT_PLUGIN_NAME
    private lateinit var crashlytics: FirebaseCrashlytics

    @NonNull
    override fun getPluginName(): String = pluginName

    @UsedByGodot
    fun init() {
        FirebaseApp.initializeApp(activity!!.applicationContext)
        crashlytics = Firebase.crashlytics
    }

    @UsedByGodot
    fun checkForUnsentReports(): Boolean = false // TODO

    @UsedByGodot
    fun deleteUnsentReports() = crashlytics.deleteUnsentReports()

    @UsedByGodot
    fun didCrashOnPreviousExecution(): Boolean = crashlytics.didCrashOnPreviousExecution()

    @UsedByGodot
    fun log(message: String) = crashlytics.log(message)

    @UsedByGodot
    fun recordException(message: String) {
        val throwable: Throwable = Throwable(message)
        crashlytics.recordException(throwable);
    }

    @UsedByGodot
    fun sendUnsentReports() = crashlytics.sendUnsentReports()

    @UsedByGodot
    fun setCrashlyticsCollectionEnabled(enabled: Boolean) =
        crashlytics.setCrashlyticsCollectionEnabled(enabled)

    @UsedByGodot
    fun setCustomKeyB(key: String, value: Boolean) = crashlytics.setCustomKey(key, value)

    @UsedByGodot
    fun setCustomKeyF(key: String, value: Float) = crashlytics.setCustomKey(key, value)

    @UsedByGodot
    fun setCustomKeyI(key: String, value: Int) = crashlytics.setCustomKey(key, value)

    @UsedByGodot
    fun setCustomKeyS(key: String, value: String) = crashlytics.setCustomKey(key, value)

    @UsedByGodot
    fun setUserId(identifier: String) = crashlytics.setUserId(identifier)

    @UsedByGodot
    fun crash(): Boolean = throw RuntimeException("From $pluginName")
}
